### PR TITLE
Update bson corpus

### PR DESCRIPTION
**Pull request by kmahar**
@mbroadst do you want me to make the travis change a separate PR? (also, do we need all the stuff in there with different Mongo versions, given that we don't actually use Mongo in this module?)

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,8 +1,34 @@
+sudo: required
```
Comment (mbroadst) [inline .travis.yml]: `sudo: false`

File: .travis.yml [hunk 2]
```diff
@@ -1,8 +1,34 @@
...
-  - "5"
-  - "6"
-  - "7"
+
+cache:
+  yarn: true
+  directories:
+    - node_modules
+
+jobs:
```
Comment (mbroadst) [inline .travis.yml]: remove the whole `jobs` section, since we don't actually need to run against mongodb, and replace it with:
```
node_js:
  - 4
  - 6
  - 8
```

File: lib/bson/decimal128.js [hunk 3]
```diff
@@ -1,6 +1,7 @@
 'use strict';
 
-var Long = require('./long');
+var Long = require('./long'),
+  f = require('util').format;
```
Comment (mbroadst) [inline lib/bson/decimal128.js]: as long as you're already here, the preferred type for top level includes is `const` since they are not meant to change.

File: lib/bson/decimal128.js [hunk 4]
```diff
@@ -79,13 +80,12 @@ var isDigit = function(value) {
...
   var DIVISOR = Long.fromNumber(1000 * 1000 * 1000);
   var _rem = Long.fromNumber(0);
-  var i = 0;
 
   if (!value.parts[0] && !value.parts[1] && !value.parts[2] && !value.parts[3]) {
     return { quotient: value, rem: _rem };
   }
 
-  for (i = 0; i <= 3; i++) {
+  for (var i = 0; i <= 3; i++) {
```
Comment (mbroadst) [inline lib/bson/decimal128.js]: use `let` here so `i` is lexically scoped to the for statement

File: lib/bson/decimal128.js [hunk 5]
```diff
@@ -245,9 +242,29 @@ Decimal128.fromString = function(string) {
     throw new Error('' + string + ' not a valid Decimal128 string');
   }
 
-  // Check if we have an illegal exponent format
-  if (stringMatch && stringMatch[4] && stringMatch[2] === undefined) {
-    throw new Error('' + string + ' not a valid Decimal128 string');
+  if (stringMatch) {
+    // full_match = stringMatch[0]
```
Comment (mbroadst) [inline lib/bson/decimal128.js]: these are for documentation purposes, not left-over commented out code, correct?

File: lib/bson/decimal128.js [hunk 6]
```diff
@@ -290,21 +307,14 @@ Decimal128.fromString = function(string) {
...
+    if (sawRadix) radixPosition = radixPosition + 1;
 
     nDigitsRead = nDigitsRead + 1;
     index = index + 1;
   }
 
-  if (sawRadix && !nDigitsRead) {
-    throw new Error('' + string + ' not a valid Decimal128 string');
-  }
+  if (sawRadix && !nDigitsRead) throw new Error('' + string + ' not a valid Decimal128 string');
```
Comment (mbroadst) [inline lib/bson/decimal128.js]: perhaps we should wrap the string in quotes?
```
throw new Error(`"${string}" not a valid Decimal128 string`);
```

Comment (mbroadst) [reply lib/bson/decimal128.js]: Also note I used template strings here, it's a convenient way to not need that imported `f` format function above.

File: lib/bson/parser/deserializer.js [hunk 7]
```diff
@@ -384,7 +392,8 @@ var deserializeObject = function(buffer, index, options, isArray) {
         buffer[index + stringSize - 1] !== 0
       )
```
Comment (mbroadst) [inline lib/bson/parser/deserializer.js]: maybe we should wrap this `throw` in braces? It's a little hard to follow

Comment (mbroadst) [inline lib/bson/parser/deserializer.js]: throw new Error('' + string + ' not a valid Decimal128 string');

File: lib/bson/parser/deserializer.js [hunk 8]
```diff
@@ -539,7 +561,7 @@ var deserializeObject = function(buffer, index, options, isArray) {
           elementType.toString(16) +
           ' for fieldname "' +
           name +
-          '", are you using the latest BSON parser'
+          '", are you using the latest BSON parser?'
```
Comment (mbroadst) [inline lib/bson/parser/deserializer.js]: nice ❓

File: lib/bson/parser/serializer.js [hunk 9]
```diff
@@ -354,7 +358,8 @@ var serializeObject = function(
   );
   // Pop stack
   path.pop();
-  // Write size
+
+  // var size = endIndex - index;
```
Comment (mbroadst) [inline lib/bson/parser/serializer.js]: cruft?

File: lib/bson/parser/serializer.js [hunk 10]
```diff
@@ -629,33 +638,15 @@ var serializeDBRef = function(buffer, key, value, index, depth, serializeFunctio
...
-        $ref: value.namespace,
-        $id: value.oid
-      },
-      false,
-      index,
-      depth + 1,
-      serializeFunctions
-    );
-  }
+  var output = {
```
Comment (mbroadst) [inline lib/bson/parser/serializer.js]: great refactor ++

File: test/node/bson_test.js [hunk 11]
```diff
@@ -1532,13 +1540,12 @@ exports['Should handle complicated all typed object'] = function(test) {
...
-  assertBuffersEqual(test, serialized_data, serialized_data2, 0);
+
+  expect(serialized_data).to.deep.equal(serialized_data2);
 
   serialized_data2 = createBSON().serialize(doc2, false, true);
 
-  for (var i = 0; i < serialized_data2.length; i++) {
-    require('assert').equal(serialized_data2[i], serialized_data[i]);
-  }
+  expect(serialized_data).to.deep.equal(serialized_data2);
```
Comment (mbroadst) [inline test/node/bson_test.js]: man whaaaat was going on in that old version of this

File: test/node/decimal128_tests.js [hunk 12]
```diff
@@ -808,29 +807,31 @@ exports['fromString exponent normalization'] = function(test) {
   );
   test.deepEqual(bytes, result.bytes);
 
+  // this should throw error according to spec.
```
Comment (mbroadst) [inline test/node/decimal128_tests.js]: and it's covered by the corpus tests?

File: lib/bson/decimal128.js (hunk 5 above)
Comment (kmahar) [reply lib/bson/decimal128.js]: yeah, I spent a while figuring out what exactly the regex was giving back, so I figured I'd leave it for future reference.

File: lib/bson/parser/deserializer.js (hunk 7 above)
Comment (kmahar) [reply lib/bson/parser/deserializer.js]: huh?

File: lib/bson/parser/deserializer.js (hunk 8 above)
Comment (kmahar) [reply lib/bson/parser/deserializer.js]: thx

File: test/node/bson_test.js (hunk 11 above)
Comment (kmahar) [reply test/node/bson_test.js]: ¯\_(ツ)_/¯

File: lib/bson/parser/deserializer.js (hunk 7 above)
Comment (mbroadst) [reply lib/bson/parser/deserializer.js]: I have no idea how this got here...

File: test/node/decimal128_tests.js (hunk 12 above)
Comment (kmahar) [reply test/node/decimal128_tests.js]: yep, one of the cases that should throw a parse error here: https://github.com/mongodb/js-bson/pull/218/files?diff=unified#diff-50fc67aa335868ab2ed1e27d8492655bR168
